### PR TITLE
add Crud trait and implementations

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

File: src/vtn/api/event.rs [hunk 1]
```diff
@@ -15,71 +13,126 @@ use openadr::wire::target::TargetLabel;
...
+                .filter_map(|(_, p)| {
+                    p.content
+                        .event_name
+                        .clone()
+                        .map(|name| (name, p.id.clone()))
+                })
+                .find(|(name, _)| name == new_event_name)
+            {
+                warn!(id=%id, event_name=%name, "Conflicting event_name");
+                return Err(AppError::Conflict(format!(
```
Comment (tdittr) [inline src/vtn/api/event.rs:43]: Event names don't need to be unique, in contrast to program names which have to be unique.

File: src/vtn/api/event.rs [hunk 2]
```diff
@@ -15,71 +13,126 @@ use openadr::wire::target::TargetLabel;
...
+    }
+
+    async fn update(&self, id: &Self::Id, content: Self::NewType) -> Result<Event, Self::Error> {
+        if let Some((_, conflict)) = self.events.write().await.iter().find(|(inner_id, p)| {
+            id != *inner_id
+                && content.event_name.is_some()
+                && p.content.event_name == content.event_name
+        }) {
+            warn!(updated=%id, conflicting=%conflict.id, event_name=?content.event_name, "Conflicting event_name");
+            return Err(AppError::Conflict(format!(
```
Comment (tdittr) [inline src/vtn/api/event.rs:89]: Does not need to be unique

**Review by tdittr — CHANGES_REQUESTED** [review]
Looks good, except that event names do not need to be unique. I wonder if there would be a good way to avoid the `<AppState as Crud<T>>::...` somehow. Maybe that would require AppState to be split up?

File: src/vtn/api/event.rs (hunk 1 above)
Comment (folkertdev) [reply src/vtn/api/event.rs:43]: ok, pretty sure I did not write that code, just moved it, but we can fix it now

**Pull request merged**